src/script/api/script_tile.hpp

```diff
@@ -158,11 +159,28 @@ class ScriptTile : public ScriptObject {
 	 */
 	static bool IsBuildableRectangle(TileIndex tile, uint width, uint height);
 
+	/**
+	 * Checks whether the given tile is actually a sea tile.
+	 * @param tile The tile to check on.
+	 * @pre ScriptMap::IsValidTile(tile).
+	 * @return True if and only if the tile is a sea tile.
+	 */
+	static bool IsSeaTile(TileIndex tile);
+
+	/**
+	 * Checks whether the given tile is actually a river tile.
+	 * @param tile The tile to check on.
+	 * @pre ScriptMap::IsValidTile(tile).
+	 * @return True if and only if the tile is a river tile.
+	 */
+	static bool IsRiverTile(TileIndex tile);
+
 	/**
 	 * Checks whether the given tile is actually a water tile.
 	 * @param tile The tile to check on.
 	 * @pre ScriptMap::IsValidTile(tile).
 	 * @return True if and only if the tile is a water tile.
+	 * @note Returns false when a buoy was placed on the tile.
```

Comment:
Nitpick, but "was placed" is wrong tense here. `Also returns false when a buoy is on the tile` would be more clear, I think? Mind fixing that real quick? :)

Reply:
What does it return for various other water-like tiles?
 * Ship depots (passable by ships)
 * Water-based industries and objects (impassable)

I was going to complain that this was an unintuitive exception before realising the function already existed.

Do the newly-added `IsSeaTile`/`IsRiverTile` have the same exception?  (looks like the answer is 'yes') If so, should probably note it there too.

There doesn't seem to be a function that checks for buoys, nor for 'passable water tiles' in general; would adding one be useful for AIs? I guess it would need a direction argument to make sense for coasts/depots and then be too complicated.

Reply:
It's hard to answer this without consulting the results. There are test cases in regression that checks the tiles posted in the screenshot above. Rivers weren't tested, the map doesn't contain any river. Oil Rig wasn't tested either.

Item | IsWaterTile | IsSeaTile | IsRiverTile | IsCanalTile | IsCoastTile
---- | ----------- | ---------- | ----------- | ----------- | -----------
Ship Depot | yes | no | no | no | no
Buoy | no | no | no | no | no | no
Sea Tile | yes | yes | no | no | no
Lock Lower Part | yes | no | no | no | no
Lock Middle Part | yes | no | no | no | no
Lock Upper Part | yes | no | no | no | no
Coast Half-Tile | no | no | no | no | yes
Coast Full-Tile | no | no | no | no | yes
Canal Tile | yes | no | no | yes | no
Grass/Tree Ground Tile | no | no | no | no | no
Dock on Water | no | no | no | no | no
Dock on Slope | no | no | no | no | no

Reply:
Thanks for the detailed table!

Looking at that, I honestly think it would be better to return `true` for buoys in IsWaterTile and consider the previous behaviour as a bug instead of documenting it. It makes no sense for ship depots to qualify but not buoys.

I might make a separate PR.